### DISCUSSION TITLE
Correct way to bulk insert rows to a table

### DISCUSSION THREAD

**Discussion by sebastian-correa** [sqlalchemy/sqlalchemy#10537] [Usage Questions] [answered]
# Summary
Using SQLAlchemy, I want to bulk insert 230k rows (8 columns) to a Postgres table. Due to an external constraint, we have to support `pg8000` and `psycopg2`, though we almost exclusively use `psycopg2`.

Depending on how I do this, the process with `pg8000` outright fails or is extremely slow. Note that I want to insert 1.84M values, which is higher than the 32767 limit Postgres has.

# Questions
1. What's the preferred way to bulk insert rows in a scenario like the example below? SQLAlchemy docs are a bit confusing to me in that I'm not sure if I should favor `insert(MyModel).values(values)` or `session.execute(insert(MyModel), values)`. It seems to me like the second is preferred.
2. Am I doing something wrong or is `pg8000` just not as performant when used in conjunction with SQLAlchemy to batch insert rows?

# Example
```python
# %%
from random import randint
from time import perf_counter

from sqlalchemy import Column, Integer, create_engine, delete
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.engine import URL
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

# %%

Base = declarative_base()


class MyModel(Base):
    __tablename__ = "my_table"

    id = Column(Integer, primary_key=True, autoincrement=True)
    column_a = Column(Integer)
    column_b = Column(Integer)
    column_c = Column(Integer)
    column_d = Column(Integer)
    column_e = Column(Integer)
    column_f = Column(Integer)
    column_g = Column(Integer)
    column_h = Column(Integer)

    column_with_server_default = Column(Integer, server_default="32")


# %%
DB_ENGINE_psycopg2 = create_engine(
    url=URL.create(
        drivername="postgresql+psycopg2",
        username="root",
        password="root",
        host="localhost",
        port=3232,
        database="test_db",
    ),
    future=False,
)
Session_pyscopg2 = sessionmaker(bind=DB_ENGINE_psycopg2)

DB_ENGINE_pg8000 = create_engine(
    url=URL.create(
        drivername="postgresql+pg8000",
        username="root",
        password="root",
        host="localhost",
        port=3232,
        database="test_db",
    ),
    future=False,
)
Session_pg8000 = sessionmaker(bind=DB_ENGINE_pg8000)

# %%
Base.metadata.create_all(DB_ENGINE_psycopg2)

with Session_pyscopg2() as session:
    session.execute(delete(MyModel))
    session.commit()

# %%
values = [
    {
        "column_a": randint(1, 200),
        "column_b": randint(1, 200),
        "column_c": randint(1, 200),
        "column_d": randint(1, 200),
        "column_e": randint(1, 200),
        "column_f": randint(1, 200),
        "column_g": randint(1, 200),
        "column_h": randint(1, 200),
    }
    for _ in range(230_000)
]

# %%
# Should work.
start = perf_counter()
with Session_pyscopg2() as session:
    statement = insert(MyModel).values(values)
    session.execute(statement)
    session.commit()

print(f"insert(MyModel).values(values) took {perf_counter() - start}s with psycopg.")

# %%
# Should fail
start = perf_counter()
with Session_pg8000() as session:
    statement = insert(MyModel).values(values)
    session.execute(statement)
    session.commit()

print(f"insert(MyModel).values(values) took {perf_counter() - start}s with pg8000.")

# %%
start = perf_counter()
with Session_pg8000() as session:
    statement = insert(MyModel)
    session.execute(statement, values)
    session.commit()

print(f"insert(MyModel) with values in execute took {perf_counter() - start}s with pg8000.")


# %%
# Should work
start = perf_counter()
with Session_pyscopg2() as session:
    statement = insert(MyModel)
    session.execute(statement, values)
    session.commit()

print(f"insert(MyModel) with values in execute took {perf_counter() - start}s with pyscopg.")
```

Results:
* Psycopg2:
    * `insert(MyModel).values(values)`: 27s.
    * `session.exec(insert(MyModel), values)`: 4.73s.
* pg8000:
    * `insert(MyModel).values(values)`: Doesn't run, hits the Postgres 32767 value limit per insert.
    * `session.exec(insert(MyModel), values)`: 406s.

# Env details:
* Python 3.10.8.
* Mac M1.
* MacOS 13.5.2 (22G91).
* Postgres on Docker (postgres:14-alpine).
* SQLAlchemy==1.4.41 (can't migrate to >2 for now due to `sqlmodel`).
* psycopg2-binary==2.9.9
* pg8000==1.30.2

# Notes
Importing `insert` from Postgres because this operation is done alongside a batch upsert that faces the same issue.

**Comment by zzzeek** [comment, marked as answer]
you would use:

    session.execute(insert(MyModel), values)

which will insert rows in batches with the psycopg2 driver specifically in 1.4.x; in 2.0, **all** postgresql drivers use the new batching.


pg8000 is written in pure python and is the least performant PostgreSQL driver , ~~which you can see where psycopg2 is doing it in less than 5 seconds, which is pretty good, vs. many minutes for pg8000, which is all overhead in the pg8000 driver.~~

since you are using 1.4, the difference in speed for psycopg2 vs pg8000 is because 1.4 has [fast execution helpers](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#psycopg2-executemany-mode) for psycopg2 only.   in 2.0, all drivers use a new approach that is roughly equivalent to these helpers so [pg8000 should be an order of magnitude faster in 2.0](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#benchmarks)

if you need to INSERT millions of rows quickly, use [COPY](https://www.postgresql.org/docs/current/sql-copy.html), for which psycopg2 and psycopg3 provides a direct API here:  https://www.psycopg.org/psycopg3/docs/basic/copy.html

**Reply by sebastian-correa** [reply]
Okay, so this low performance in `pg8000` is expected and not SQLAlchemy-user error on my side? If that's the case, I can do `session.execute(insert(MyModel), values)` everywhere and live with the slow inserts 🤷‍♂️.

**Reply by zzzeek** [reply]
~~no~~ sort of, I had to edit my answer when I saw you were on 1.4.  read the updated text

**Reply by zzzeek** [reply]
i mean yes, pg8000 is a lot slower anyway, but not by two orders of magnitude for equivalent operations

**Reply by sebastian-correa** [reply]
Gotcha. I'll migrate to using `session.execute(insert(MyModel), values)` to keep support for now (had resorted to batching by hand with a `for` loop :S) and will try to move to `SQLAlchemy` >2 soon.

Might have to drop `sqlmodel`, but that's out of scope for `SQLAlchemy`. I saw that there's some news in `SQLAlchemy` >2 around type-hints, so `sqlmodel` might become unnecessary in our codebase. I'll take a look.

Thanks!

**Reply by zzzeek** [reply]
I'm really floored that SQLModel hasn't added SQLA 2 support yet, there is AFAIK no technical barrier to this and it's a brand new library.  we're coming up on a full year since 2.0 was released

**Reply by sebastian-correa** [reply]
They [plan](https://github.com/tiangolo/sqlmodel/issues/654) on adding support. My guess is that tiangolo has been busy with FastAPI and other bigger projects...

Your input over there might help change the order of the roadmap :).

**Reply by CaselIT** [reply]
It never seemed very supported as a library sqlmodel. It's a shame since a lot of people do like pydantic

regarding the question, chiming in by mentioning that you may be able to use copy also with pg8000 https://github.com/tlocke/pg8000#pg8000dbapicursorexecuteoperation-argsnone-streamnone

**Reply by sebastian-correa** [reply]
> It's a shame since a lot of people do like pydantic

Yes, type hints and pydantic are the 2 main features I like.

I'll look into the `COPY` statement. Thanks to both!

**Reply by jmorese** [reply]
Hi @zzzeek , @CaselIT 

I am facing the exact same case as @sebastian-correa but in a select statement. 

In my case we are doing a query like follows:

```python

very_big_list = [ randint(1, 200) for _ in range(230_000)]

with Session_pg8000() as session:
    statement = select(MyModel).where(MyModel.column_a.in_(very_big_list))
    session.execute(statement)
```

The error I am getting is `struct.error: 'h' format requires -32768 <= number <= 32767`

Is there something I can do similar to what you suggested with the `insert` statement?

**Reply by zzzeek** [reply]
that's just the driver + db limitation, you need to break up that IN list into segments and do multiple SELECT statements most ideally, or do WHERE COL IN (...) OR COL IN (...) .... which would also be fairly wasteful, or the most high performing way depending on requirements, put the values into a temporary table and JOIN to that table instead.

The error here is the driver, I can reproduce here:

```
 File "/home/classic/.venv3/lib64/python3.12/site-packages/pg8000/legacy.py", line 254, in execute
    self._context = self._c.execute_unnamed(
                    ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/classic/.venv3/lib64/python3.12/site-packages/pg8000/core.py", line 701, in execute_unnamed
    self.send_BIND(NULL_BYTE, params)
  File "/home/classic/.venv3/lib64/python3.12/site-packages/pg8000/core.py", line 765, in send_BIND
    NULL_BYTE + statement_name_bin + h_pack(0) + h_pack(len(params))
                                                 ^^^^^^^^^^^^^^^^^^^

```

note the package where the error occurs

whereas reducing the number of elements has no issue

**Reply by jmorese** [reply]
I ended up splitting the select statement into different segments. Thank you very much for your prompt reply!

**Reply by nicksawhney** [reply]
What if we want to have a function that does `on_conflict_do_update` do batching? 

```
            insert_stmt = insert(table.table).values(list(data_iter))
            upsert_stmt = insert_stmt.on_conflict_do_update(
                index_elements=self.indexes,
                set_={k: getattr(insert_stmt.excluded, k) for k in keys},
            )
            conn.execute(upsert_stmt)
```

**Reply by sebastian-correa** [reply]
Hey @nicksawhney. Might be a bit late for you now, but it's the same thing! You pass the elements as parameters, like this:

```python
            insert_stmt = insert(SomeModel)
            upsert_stmt = insert_stmt.on_conflict_do_update(
                index_elements=...,
                set_={k: getattr(insert_stmt.excluded, k) for k in keys},
            )
            conn.execute(upsert_stmt, parameters=data_iter)
```

assuming `data_iter` has `dict[str, Any]` complying with `SomeModel`. I'm not familiar with internals, but I'd advice against using `.values`.

Edit: BTW, we've moved away form `pg8000`, so I can't comment on how to circumvent the parameter limit.
